File: aiwos/aiwos-backend/app/schemas/analytics.py
import uuid
from datetime import datetime
from pydantic import BaseModel


# ── Execution Metrics (new endpoint) ─────────────────────────────────────────

class WorkflowExecutionMetrics(BaseModel):
    total_executions: int
    completed: int
    failed: int
    running: int
    success_rate: float
    avg_duration_seconds: float


class ProviderUsageStat(BaseModel):
    provider: str
    count: int
    percentage: float


class AgentUtilizationStat(BaseModel):
    agent_id: str
    agent_name: str
    total_executions: int
    successful: int
    failed: int
    success_rate: float
    avg_duration_ms: float


class ExecutionMetricsResponse(BaseModel):
    total_executions: int
    successful_executions: int
    failed_executions: int
    retry_count: int
    avg_duration_ms: float
    success_rate: float
    workflow_metrics: WorkflowExecutionMetrics
    provider_usage: list[ProviderUsageStat]
    agent_utilization: list[AgentUtilizationStat]


# ── Analytics Metrics (dedicated analytics page) ──────────────────────────────

class TrendDataPoint(BaseModel):
    date: str
    created: int
    completed: int


class DepartmentTaskStat(BaseModel):
    department: str
    tasks: int
    percentage: float


class AnalyticsMetricsResponse(BaseModel):
    # Task metrics
    total_tasks: int
    completed_tasks: int
    in_progress_tasks: int
    pending_tasks: int
    failed_tasks: int
    completion_rate: float

    # Execution metrics
    total_executions: int
    successful_executions: int
    failed_executions: int
    retried_executions: int
    avg_response_time_seconds: float
    success_rate: float

    # Agent metrics
    total_agents: int
    active_agents: int

    # Workflow / project metrics
    total_workflows: int
    total_projects: int
    active_projects: int
    completed_projects: int

    # Period-over-period % changes (current 7 days vs previous 7 days)
    tasks_change_pct: float
    completed_change_pct: float
    success_rate_change_pct: float
    response_time_change_pct: float  # negative = response time improved

    # Chart data
    task_completion_trend: list[TrendDataPoint]
    tasks_by_department: list[DepartmentTaskStat]


# ── Dashboard (existing) ───────────────────────────────────────────────────────

class DashboardStats(BaseModel):
    total_agents: int
    running_tasks: int
    tasks_completed: int
    total_projects: int
    active_projects: int
    total_executions: int
    workflow_count: int
    total_cost_today: float
    # Execution health breakdown
    executions_successful: int = 0
    executions_failed: int = 0
    executions_retried: int = 0
    executions_with_fallback: int = 0


class DepartmentStat(BaseModel):
    id: str
    name: str
    agent_count: int
    completed_tasks: int
    total_tasks: int


class ActivityItem(BaseModel):
    id: str
    agent_name: str | None
    task_title: str | None
    action: str
    timestamp: datetime
    status: str


class TopAgentStat(BaseModel):
    id: str
    name: str
    tasks_completed: int
    tasks_failed: int
    success_rate: float


class WeeklyCompletion(BaseModel):
    date: str
    completed: int


class DashboardResponse(BaseModel):
    stats: DashboardStats
    departments: list[DepartmentStat]
    recent_activities: list[ActivityItem]
    top_agents: list[TopAgentStat]
    weekly_completions: list[WeeklyCompletion]

File: aiwos/lib/api/analytics.ts
import { apiClient } from "./client";

export type DashboardStats = {
  total_agents: number;
  running_tasks: number;
  tasks_completed: number;
  total_projects: number;
  active_projects: number;
  total_executions: number;
  workflow_count: number;
  total_cost_today: number;
  // Execution health breakdown
  executions_successful: number;
  executions_failed: number;
  executions_retried: number;
  executions_with_fallback: number;
};

export type DepartmentStat = {
  id: string;
  name: string;
  agent_count: number;
  completed_tasks: number;
  total_tasks: number;
};

export type ActivityItem = {
  id: string;
  agent_name: string | null;
  task_title: string | null;
  action: string;
  timestamp: string;
  status: string;
};

export type TopAgentStat = {
  id: string;
  name: string;
  tasks_completed: number;
  tasks_failed: number;
  success_rate: number;
};

export type WeeklyCompletion = {
  date: string;
  completed: number;
};

export type DashboardResponse = {
  stats: DashboardStats;
  departments: DepartmentStat[];
  recent_activities: ActivityItem[];
  top_agents: TopAgentStat[];
  weekly_completions: WeeklyCompletion[];
};

export const analyticsApi = {
  dashboard: (organization_id: string) =>
    apiClient
      .get<DashboardResponse>("/analytics/dashboard", {
        params: { organization_id },
      })
      .then((r) => r.data),

  metrics: (organization_id: string, time_range: string) =>
    apiClient
      .get<AnalyticsMetricsResponse>("/analytics/metrics", {
        params: { organization_id, time_range },
      })
      .then((r) => r.data),

  executionMetrics: (organization_id: string) =>
    apiClient
      .get<ExecutionMetricsResponse>("/analytics/execution-metrics", {
        params: { organization_id },
      })
      .then((r) => r.data),
};

// ── Analytics Metrics types ───────────────────────────────────────────────────

export type TrendDataPoint = {
  date: string;
  created: number;
  completed: number;
};

export type DepartmentTaskStat = {
  department: string;
  tasks: number;
  percentage: number;
};

// ── Execution Metrics ─────────────────────────────────────────────────────────

export type WorkflowExecutionMetrics = {
  total_executions: number;
  completed: number;
  failed: number;
  running: number;
  success_rate: number;
  avg_duration_seconds: number;
};

export type ProviderUsageStat = {
  provider: string;
  count: number;
  percentage: number;
};

export type AgentUtilizationStat = {
  agent_id: string;
  agent_name: string;
  total_executions: number;
  successful: number;
  failed: number;
  success_rate: number;
  avg_duration_ms: number;
};

export type ExecutionMetricsResponse = {
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  retry_count: number;
  avg_duration_ms: number;
  success_rate: number;
  workflow_metrics: WorkflowExecutionMetrics;
  provider_usage: ProviderUsageStat[];
  agent_utilization: AgentUtilizationStat[];
};

export type AnalyticsMetricsResponse = {
  // Task metrics
  total_tasks: number;
  completed_tasks: number;
  in_progress_tasks: number;
  pending_tasks: number;
  failed_tasks: number;
  completion_rate: number;

  // Execution metrics
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  retried_executions: number;
  avg_response_time_seconds: number;
  success_rate: number;

  // Agent metrics
  total_agents: number;
  active_agents: number;

  // Workflow / project metrics
  total_workflows: number;
  total_projects: number;
  active_projects: number;
  completed_projects: number;

  // Period-over-period % changes
  tasks_change_pct: number;
  completed_change_pct: number;
  success_rate_change_pct: number;
  response_time_change_pct: number;

  // Chart data
  task_completion_trend: TrendDataPoint[];
  tasks_by_department: DepartmentTaskStat[];
};
